pallets/gear-voucher/src/lib.rs

```diff
@@ -100,151 +98,413 @@ pub mod pallet {
-        /// to `program_id` program.
-        ///
-        /// The dispatch origin for this call must be _Signed_.
+    impl<T: Config> Pallet<T>
+    where
+        T::AccountId: Origin,
+    {
+        /// Issue a new voucher.
         ///
-        /// - `to`: The voucher holder account id.
-        /// - `program`: The program id, messages to whom can be paid with the voucher.
-        /// NOTE: the fact a program with such id exists in storage is not checked - it's
-        /// a caller's responsibility to ensure the consistency of the input parameters.
-        /// - `amount`: The voucher amount.
+        /// Deposits event `VoucherIssued`, that contains `VoucherId` to be
+        /// used by spender for balance-less on-chain interactions.
         ///
-        /// ## Complexity
-        /// O(Z + C) where Z is the length of the call and C its execution weight.
+        /// Arguments:
+        /// * spender:  user id that is eligible to use the voucher;
+        /// * balance:  voucher balance could be used for tx fees and gas,
+        /// * programs: pool of programs spender can interact with,
+        ///             if None - means any program,
+        ///             limited by Config param;
+        /// * validity: amount of blocks voucher could be used by spender
+        ///             and couldn't be revoked by owner,
+        ///             couldn't be zero.
         #[pallet::call_index(0)]
-        #[pallet::weight(<T as Config>::WeightInfo::issue())]
+        #[pallet::weight(T::WeightInfo::issue())]
         pub fn issue(
             origin: OriginFor<T>,
-            to: AccountIdLookupOf<T>,
-            program: ProgramId,
-            value: BalanceOf<T>,
+            spender: AccountIdOf<T>,
+            balance: BalanceOf<T>,
+            programs: Option<BTreeSet<ProgramId>>,
+            validity: BlockNumberFor<T>,
         ) -> DispatchResultWithPostInfo {
-            let who = ensure_signed(origin)?;
-            let to = T::Lookup::lookup(to)?;
+            // Ensuring origin.
+            let owner = ensure_signed(origin)?;
+
+            // Asserting amount of programs.
+            if let Some(ref programs) = programs {
+                ensure!(
+                    programs.len() <= T::MaxProgramsAmount::get().saturated_into(),
+                    Error::<T>::MaxProgramsLimitExceeded
+                )
+            }
+
+            // Creating voucher id.
+            let voucher_id = VoucherId::generate::<T>();
+
+            // Transferring funds to voucher account.
+            T::Currency::transfer(
+                &owner,
+                &voucher_id.cast(),
+                balance,
+                ExistenceRequirement::KeepAlive,
+            )
+            .map_err(|_| Error::<T>::BalanceTransfer)?;
+
+            // Asserting non-zero validity.
+            ensure!(!validity.is_zero(), Error::<T>::ZeroValidity);
```

Comment:
Better place this check before any access to storage, before asserting amount of programs or after for example.